[MLIR][Linalg] Scalable Vectorization of Reduction on the Trailing Dimension

zhaoshiz opened the pull request:
Allow scalable vectorization of linalg::reduce and linalg::generic with reduction iterator with two restrictions:
1. The reduction dim is the last (innermost) dim of the op; and
2. Only the reduction dim is requested for scalable vectorization.

One exception is that scalable vectorization of the reduction dim in Matmul-like ops are not supported even above restrictions are met.

    Allowed combinations of scalable flags and iterator types:
    Matmul:
         Iterators: ["parallel", "parallel", "reduction"]
    Scalable Flags: ["true",     "true",     "false"]
                    ["false",    "true",     "false"]
    Matvec:
         Iterators: ["parallel", "reduction"]
    Scalable Flags: ["false",    "true"]
                    ["true",     "false"]

zhaoshiz commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1942,13 +1951,30 @@ vectorizeScalableVectorPrecondition(Operation *op,
+    return success();
+  }
+
   bool isScalable = inputScalableVecDims.back();
```
this seems to be missed by 5f6c03636d22029550c8ab94388827e728a0c5fe, lifting the restriction that only the trailing dimension can be scalably vectorized.

@banach-space, maybe we can remove these lines?

banach-space replied:
Indeed, feel free to remove (just add a note in the summary).

zhaoshiz commented on mlir/test/Integration/Dialect/Linalg/CPU/ArmSVE/generic_reduce_2d.mlir:
```diff
@@ -0,0 +1,95 @@
+// DEFINE:    -convert-vector-to-llvm="enable-arm-sve" -test-lower-to-llvm -o %t
+// DEFINE: %{entry_point} = generic_reduce_2d_f32
+// DEFINE: %{run} = %mcr_aarch64_cmd %t -e %{entry_point} -entry-point-result=void --march=aarch64 --mattr="+sve"\
+// DEFINE:    -shared-libs=%mlir_native_utils_lib_dir/libmlir_runner_utils%shlibext,%mlir_native_utils_lib_dir/libmlir_c_runner_utils%shlibext
```
`%mlir_native_utils_lib_dir/libmlir_runner_utils%shlibext` will expands to `config.arm_emulator_utils_lib_dir/libmlir_runner_utils%shlibext`.

config.arm_emulator_utils_lib_dir is defined in https://github.com/llvm/llvm-project/blob/main/mlir/test/lit.site.cfg.py.in#L59

banach-space commented:
@zhaoshiz Apologies for the delay with this - I was travelling last week and still catching up with PRs. If not today, I promise to go over this tomorrow. In the meantime, would you mind fixing the conflict?

zhaoshiz commented:
> @zhaoshiz Apologies for the delay with this - I was travelling last week and still catching up with PRs. If not today, I promise to go over this tomorrow. In the meantime, would you mind fixing the conflict?

No worries @banach-space. By "conflict" do you mean
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1950C1-L1952C22?

I tried to change it locally to check that all dims are not scalable:
`
  if (llvm::none_of(inputScalableVecDims,
                    [](bool isScalable) { return isScalable; }))
    return success();
`
but broke the the integration test https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/Dialect/Linalg/CPU/ArmSVE/matmul.mlir.

Essentially we are doing white-list of linalg ops in function `vectorizeScalableVectorPrecondition`, the existing check
`  bool isScalable = inputScalableVecDims.back();
  if (!isScalable)
    return success();` allows vector sizes like [[4],4] or [1, [4], 1] to proceed regardless of the linalg op being vectorized. So to fix it we'll need to verifiy a lot of ops (e.g.: matvec) can be scalably vectorized and add them.

I have another question: a great number of mlir integration tests are written with `-shared-libs=%mlir_runner_utils,%mlir_c_runner_utils` which looks for the shared libs in <build_dir/lib> on the host machine (in my case, x86_64-linux). These tests will fail if config-ed to run in an emulator (qemu-aarch64).
I wrote the tests in this PR with `-shared-libs=%mlir_native_utils_lib_dir/libmlir_runner_utils%shlibext,%mlir_native_utils_lib_dir/libmlir_c_runner_utils%shlibext` so they run successfully in qemu. Can you help run them on a native arm64-linux host with sve support?

banach-space commented:
> > @zhaoshiz Apologies for the delay with this - I was travelling last week and still catching up with PRs. If not today, I promise to go over this tomorrow. In the meantime, would you mind fixing the conflict?
> 
> No worries @banach-space. By "conflict" do you mean 

I meant this:

![Screenshot 2024-07-10 at 20 59 03](https://github.com/llvm/llvm-project/assets/2990321/699938c2-1386-4509-b9ba-8bf6614e2b49)

I will reply more tomorrow :)

zhaoshiz commented:
rebased and fixed the conflict

banach-space commented on mlir/test/Integration/Dialect/Linalg/CPU/ArmSVE/reduce_2d.mlir:
```diff
@@ -0,0 +1,91 @@
+
+// RUN: %{run} | FileCheck %s --check-prefix=F32
+
+func.func @reduce_2d_f32() {
```
Would it be possible to move this to "generic_reduce_2d.mlir". It feels that we should be able to re-use the TD logic.

zhaoshiz replied:
merged generic_reduce_2d.mlir into this file

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1947,13 +1956,30 @@ vectorizeScalableVectorPrecondition(Operation *op,
   // Only element-wise and 1d depthwise conv ops supported in the presence of
   // scalable dims.
-  auto linalgOp = dyn_cast<LinalgOp>(op);
   return success(linalgOp && (isElementwise(linalgOp) ||
                               isa<linalg::DepthwiseConv1DNwcWcOp>(op)));
```
This requires updating.

zhaoshiz replied:
reduce ops are checked above by new lines L1961 ~ L1975

banach-space reviewed (commented):
Thank you for working on this! 🙏🏻 Overall LG. I've made a few suggestions, but nothing major.

> Essentially we are doing white-list of linalg ops in function vectorizeScalableVectorPrecondition, the existing check
  bool isScalable = inputScalableVecDims.back(); if (!isScalable) return success(); allows vector sizes like [[4],4] or [1, [4], 1] to proceed regardless of the linalg op being vectorized. So to fix it we'll need to verifiy a lot of ops (e.g.: matvec) can be scalably vectorized and add them.
  
Please don't forget about this check: https://github.com/llvm/llvm-project/blob/1ed84a862f9ce3c60251968f23a5405f06458975/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1952-L1953

Also, note:
https://github.com/llvm/llvm-project/blob/1ed84a862f9ce3c60251968f23a5405f06458975/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1756-L1765

So, it looks like there's at least two hooks to check the preconditions for `linalg.reduce`:
* `vectorizeScalableVectorPrecondition`
* `vectorizeDynamicLinalgOpPrecondition`
* ` vectorizeLinalgOpPrecondition`

Given that we are adding limitations specific to scalable vectors, I think that `vectorizeScalableVectorPrecondition` is the right place for now.

> I have another question: 

Let me try something and I will get back to you tomorrow!

zhaoshiz commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1947,13 +1956,30 @@ vectorizeScalableVectorPrecondition(Operation *op,
+           "Number of iterator types and input scalable dims mismatch");
+    // For now, only support scalable vectorization of a reduction on the
+    // trailing dim.
+    for (size_t i = 0; i < inputScalableVecDims.size() - 1; ++i) {
+      if (inputScalableVecDims[i] && isReductionIterator(iteratorTypes[i])) {
+        LDBG("Non-trailing reduction dim requested for scalable "
+             "vectorization\n");
+        return failure();
+      }
+    }
+    return success();
+  }
```
this would work for linalg.reduce ops but prevent vectorizing dimensions with parallel iterators of linalg.generics ops, e.g.: requested vector sizes are [4, [4], 1] for the op below:
```
%result = linalg.generic {
  indexing_maps = [affine_map<(i, j, k) -> (i, k)>,
                   affine_map<(i, j, k) -> (k, j)>,
                   affine_map<(i, j, k) -> (i, j)>],
  iterator_types = ["parallel", "parallel", "reduction"]
} ins(%lhs, %rhs : tensor<8x10xf32>,tensor<10x16xf32>)
  outs(%init :tensor<8x16xf32>) {
^bb0(%lhs_one: f32, %rhs_one: f32, %init_one: f32):
  %0 = arith.mulf %lhs_one, %rhs_one : f32
  %1 = arith.addf %init_one, %0 : f32
  linalg.yield %1 : f32
} -> tensor<8x16xf32>
```

zhaoshiz commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1947,13 +1956,30 @@ vectorizeScalableVectorPrecondition(Operation *op,
     return success();
 
+  auto linalgOp = dyn_cast<LinalgOp>(op);
+  if (linalgOp && isLinalgReduction(linalgOp)) {
```
Note `isLinalgReduction(linalgOp)`

zhaoshiz commented on mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:
```diff
@@ -1778,6 +1784,9 @@ vectorizeDynamicLinalgOpPrecondition(linalg::LinalgOp op,
   if (isa<ConvolutionOpInterface>(op.getOperation()))
     return vectorizeDynamicConvOpPrecondition(op, flatten1DDepthwiseConv);
 
+  if (isLinalgReduction(op))
+    return reductionPreconditions(op);
+
```
linalg.reduce ops will fail the next check L1792~L1795 and cause vectorizeLinalgOpPrecondition() to return failure.
For static-shaped reduce ops I don't think we need it. But after tiling with scalable vector like [[4]], we got dynamic-shaped ops.

zhaoshiz reviewed (commented):
working on tidy-up the test cases, commented inline about issues in mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp

banach-space reviewed (commented):
I've finally had a bit more time for this. I think that a lot of complexities in this PR stem from the fact that `vectorizeScalableVectorPrecondition` is a bit messy and hard to extend (my fault!). Also:

> Simply removing it will break some useful cases like matmul. Making sure we allow all correct combinations of vector sizes and op types and prevent unsupported cases is beyond the scope of this PR. I'm happy to work on it later.

I think that we should refactor things a bit first and then build this PR on top of that. To quickly unblock you, here's what I'm proposing:
* https://github.com/llvm/llvm-project/pull/98639

I think that you should be able to enable "reductions" quite easily. Sadly re-basing won't be straightforward :(

Also:
> [MLIR][Linalg] Scalable Vectorization of Reduction

To me this says that you are adding scalable vectorisation of e.g. `linalg.reduce`, but in practice you are doing something more generic - allowing reduction dimensions to be scalable. It's worth updating the summary.

banach-space commented:
> I have another question: a great number of mlir integration tests are written with -shared-libs=%mlir_runner_utils,%mlir_c_runner_utils which looks for the shared libs in <build_dir/lib> on the host machine (in my case, x86_64-linux). These tests will fail if config-ed to run in an emulator (qemu-aarch64).
I wrote the tests in this PR with -shared-libs=%mlir_native_utils_lib_dir/libmlir_runner_utils%shlibext,%mlir_native_utils_lib_dir/libmlir_c_runner_utils%shlibext so they run successfully in qemu. Can you help run them on a native arm64-linux host with sve support?

Sorry for not getting back to you earlier. Could you clarify - are you asking how to run this on x86_64 or AArch64? The former will require cross-compiling. The latter, should just work ™️ 😅 If it didn't, could you share more details?

zhaoshiz commented:
> Sorry for not getting back to you earlier. Could you clarify - are you asking how to run this on x86_64 or AArch64? The former will require cross-compiling. The latter, should just work ™️ 😅 If it didn't, could you share more details?

We should be able to run the integration tests on SVE/SME in both ways: 1. qemu-aarch64 on x86_64; 2 native aarch64-linux
For 2, I'm assuming they just work with `-shared-libs=%libmlir_runner_utils,...`.
For 1, I can build llvm/mlir on x86_64 with addtional cmake variables specifying pre-built aarch64 mlir-cpu-runner and shared libs:
```
  -DARM_EMULATOR_EXECUTABLE="<path_to_qemu_bin>/qemu-aarch64" \
  -DARM_EMULATOR_OPTIONS="-L /usr/aarch64-linux-gnu" \
  -DARM_EMULATOR_MLIR_CPU_RUNNER_EXECUTABLE="<path_to_llvm_arm64_build>/bin/mlir-cpu-runner-arm64" \
  -DARM_EMULATOR_UTILS_LIB_DIR="<path_to_llvm_arm64_build>/lib"
```
and write integration tests with `-shared-libs=%mlir_native_utils_lib_dir/libmlir_runner_utils%shlibext,...` which expands to: `-shared-libs=config.arm_emulator_utils_lib_dir/libmlir_runner_utils.so,...` where "config.arm_emulator_utils_lib_dir" is set by "-DARM_EMULATOR_UTILS_LIB_DIR".

My question is will `-shared-libs=%mlir_native_utils_lib_dir/libmlir_runner_utils%shlibext,...` work in native aarch64-linux environment?

banach-space commented:
> My question is will -shared-libs=%mlir_native_utils_lib_dir/libmlir_runner_utils%shlibext,... work in native aarch64-linux environment?

It should. When `ARM_EMULATOR_UTILS_LIB_DIR` is _not set_, it's basically `config.mlir_lib_dir`: https://github.com/llvm/llvm-project/blob/93d7d9bfd4aede19dda0ebaf8aead12c2adbd13b/mlir/test/Integration/lit.local.cfg#L18-L19

The naming is not great though ... Btw, if you are building on X86, do other SVE integration tests work for you? I imagine that they fail - we've not really used `%mlir_native_utils_lib_dir` much.

zhaoshiz commented:
> if you are building on X86, do other SVE integration tests work for you?

no, all SVE test written with `%libmlir_runner_utils` will fail when built with -DARM_EMULATOR_EXECUTABLE, etc.
so to make them work in both ways, we need to update the tests to use `%mlir_native_utils_lib_dir`?

banach-space commented:
> no, all SVE test written with `%libmlir_runner_utils` will fail when built with -DARM_EMULATOR_EXECUTABLE, etc. so to make them work in both ways, we need to update the tests to use `%mlir_native_utils_lib_dir`?

Sorry about this - thank you for checking and for reporting 🙏🏻 Yes, we need to fix this. Would you have the cycles for this?

zhaoshiz commented:
> Sorry about this - thank you for checking and for reporting 🙏🏻 Yes, we need to fix this. Would you have the cycles for this?

yea I'm happy to fix that but it'll take some time for me to set up an environment to test `%mlir_native_utils_lib_dir` in native mode.

banach-space commented:
> > Sorry about this - thank you for checking and for reporting 🙏🏻 Yes, we need to fix this. Would you have the cycles for this?
> 
> yea I'm happy to fix that but it'll take some time for me to set up an environment to test `%mlir_native_utils_lib_dir` in native mode.

If you test it on X86, then I can take care of testing on AArch64. IIUC, the former is easier for you to set-up?

zhaoshiz committed:
[MLIR][Linalg] Scalable Vectorization of Reduction on the Trailing Dimension

Allow scalable vectorization of linalg::reduce and linalg::generic
with reduction iterator. For now, only reduction on the trailing
dimension is supported.

zhaoshiz committed:
[MLIR][Linalg] Add integration tests of scalable vectorization of reduction

Note: I don't have a setup to run these tests natively (arm64-linux with sve).
I am able to run them using QEMU on a x86_64-linux with below cmake variables
when building llvm:

  -DARM_EMULATOR_EXECUTABLE="<path_to_qemu_bin>/qemu-aarch64" \
  -DARM_EMULATOR_OPTIONS="-L /usr/aarch64-linux-gnu" \
  -DARM_EMULATOR_MLIR_CPU_RUNNER_EXECUTABLE="<path_to_llvm_arm64_build>/bin/mlir-cpu-runner-arm64" \
  -DARM_EMULATOR_UTILS_LIB_DIR="<path_to_llvm_arm64_build>/lib"

zhaoshiz commented:
rebased/reworked after https://github.com/llvm/llvm-project/pull/98639 is merged

zhaoshiz committed:
fix per clang-format

banach-space commented on mlir/test/Dialect/Linalg/vectorization-scalable.mlir:
```diff
@@ -189,3 +189,85 @@ module attributes {transform.with_named_sequence} {
+// CHECK:          %[[VAL_9:.*]] = vector.mask %[[VAL_4]] { vector.multi_reduction <add>, %[[VAL_5]], %[[VAL_8]] [0] : vector<[4]xf32> to f32 } : vector<[4]xi1> -> f32
+// CHECK:          %[[VAL_10:.*]] = vector.broadcast %[[VAL_9]] : f32 to vector<f32>
+// CHECK:          %[[VAL_11:.*]] = vector.transfer_write %[[VAL_10]], %[[ARG_1]][] : vector<f32>, tensor<f32>
+// CHECK:          return %[[VAL_11]] : tensor<f32>
+// CHECK:        }
```
These two lines can be skipped:
* https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices

banach-space commented on mlir/test/Dialect/Linalg/vectorization-scalable.mlir:
```diff
@@ -189,3 +189,85 @@ module attributes {transform.with_named_sequence} {
+
+// CHECK-LABEL:  func.func @vectorize_dynamic_reduction_scalable_1d(
+// CHECK-SAME:     %[[ARG_0:.*]]: tensor<?xf32>, %[[ARG_1:.*]]: tensor<f32>) -> tensor<f32> {
+// CHECK:          %[[VAL_0:.*]] = arith.constant 0 : index
+// CHECK:          %[[VAL_1:.*]] = tensor.dim %[[ARG_0]], %[[VAL_0]] : tensor<?xf32>
+// CHECK:          %[[VAL_2:.*]] = arith.constant 0 : index
+// CHECK:          %[[VAL_3:.*]] = arith.constant 0.000000e+00 : f32
+// CHECK:          %[[VAL_4:.*]] = vector.create_mask %[[VAL_1]] : vector<[4]xi1>
+// CHECK:          %[[VAL_5:.*]] = vector.mask %[[VAL_4]] { vector.transfer_read %[[ARG_0]][%[[VAL_2]]], %[[VAL_3]] {in_bounds = [true]} : tensor<?xf32>, vector<[4]xf32> } : vector<[4]xi1> -> vector<[4]xf32>
+// CHECK:          %[[VAL_6:.*]] = arith.constant 0.000000e+00 : f32
+// CHECK:          %[[VAL_7:.*]] = vector.transfer_read %[[ARG_1]][], %[[VAL_6]] : tensor<f32>, vector<f32>
+// CHECK:          %[[VAL_8:.*]] = vector.extractelement %[[VAL_7]][] : vector<f32>
+// CHECK:          %[[VAL_9:.*]] = vector.mask %[[VAL_4]] { vector.multi_reduction <add>, %[[VAL_5]], %[[VAL_8]] [0] : vector<[4]xf32> to f32 } : vector<[4]xi1> -> f32
+// CHECK:          %[[VAL_10:.*]] = vector.broadcast %[[VAL_9]] : f32 to vector<f32>
+// CHECK:          %[[VAL_11:.*]] = vector.transfer_write %[[VAL_10]], %[[ARG_1]][] : vector<f32>, tensor<f32>
```
[nit] Kind request - descriptive LIT variable names.

banach-space commented on mlir/test/Dialect/Linalg/vectorization-scalable.mlir:
```diff
@@ -189,3 +189,85 @@ module attributes {transform.with_named_sequence} {
+module attributes {transform.with_named_sequence} {
+  transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
+    %0 = transform.structured.match ops{["linalg.generic"]} in %arg1 : (!transform.any_op) -> !transform.any_op
+    transform.structured.vectorize %0 vector_sizes [1, [4]] : !transform.any_op
```
[nit] Use `[4, [8]]` instead for consistency: https://github.com/llvm/llvm-project/blob/867ff2d4268ca7eed89a24f100b67b68f5329439/mlir/test/Dialect/Linalg/vectorization.mlir#L246

banach-space commented on mlir/test/Integration/Dialect/Linalg/CPU/ArmSVE/reduce_1d.mlir:
```diff
@@ -0,0 +1,134 @@
+// REDEFINE: %{entry_point} = generic_reduce_1d_f32
+// RUN: %{run} | FileCheck %s --check-prefix=GENERIC
+
+func.func @reduce_1d_f32() {
```
Would you mind adding a test for `i16` as well? Or any integer value. Just to make sure that we check both FP and integers. Thanks!

zhaoshiz committed:
Addressed review comments

zhaoshiz committed:
Handle Matmul and Matvec's reduction dim on scalable vectorization

In summary:
1. Do not allow scalable vectorization of the reduction dim of Matmul-like ops.
2. Allow scalable vectorization on only one dim of Matvec op.

Allowed combinations of scalable flags and iterator types:
Matmul:
     Iterators: ["parallel", "parallel", "reduction"]
Scalable Flags: ["true",     "true",     "false"]
                ["false",    "true",     "false"]
Matvec:
     Iterators: ["parallel", "reduction"]
Scalable Flags: ["false",    "true"]
                ["true",     "false"]

zhaoshiz reviewed (commented):
revised per comments

zhaoshiz committed:
update per clang-format

zhaoshiz committed:
2nd update per clang-format

zhaoshiz committed:
another update per clang-format

banach-space reviewed (approved):
LGTM

Thank you for bearing with me and for addressing my comments. Great test coverage!
